### PR TITLE
Multi exec

### PR DESCRIPTION
**Pull request by DvirDukhan**
added multi-exec support

### CODE REVIEW

File: src/main/java/com/redislabs/redisgraph/impl/JRedisGraphTransaction.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+     * Execute a Cypher query with arguments
+     *
+     * @param graphId a graph to perform the query on
+     * @param query Cypher query
+     * @param args
+     * @return response with a result set
+     */
+    public Response<ResultSet> graphQuery(String graphId, String query, Object ...args){
+        String preparedQuery = Utils.prepareQuery(query, args);
+        graphCaches.putIfAbsent(graphId, new GraphCache(graphId, redisGraph));
```
Comment (gkorland) [inline src/main/java/com/redislabs/redisgraph/impl/JRedisGraphTransaction.java]: @DvirDukhan I think you should avoid creating GraphCache on each query

File: src/main/java/com/redislabs/redisgraph/impl/JRedisGraphTransaction.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        client.sendCommand(Command.QUERY, graphId, preparedQuery, "--COMPACT");
+        return getResponse(new Builder<ResultSet>() {
+            @Override
```
Comment (gkorland) [inline src/main/java/com/redislabs/redisgraph/impl/JRedisGraphTransaction.java]: Do you really need to create a new builder on each call, or can you reuse a "shared" builder?

File: src/main/java/com/redislabs/redisgraph/impl/JRedisGraphTransaction.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+
+    /**
+     * Invokes stored procedures without arguments, in multi/exec context
+     * @param graphId a graph to perform the query on
+     * @param procedure procedure name to invoke
+     * @return response with result set with the procedure data
+     */
+    public Response<ResultSet> graphCallProcedure(String graphId, String procedure){
+        return graphCallProcedure(graphId, procedure, new ArrayList<>(), new HashMap<>());
```
Comment (gkorland) [inline src/main/java/com/redislabs/redisgraph/impl/JRedisGraphTransaction.java]: I think you should set the ArrayList(0), HashMap(0)

Comment (gkorland) [reply src/main/java/com/redislabs/redisgraph/impl/JRedisGraphTransaction.java]: or can you reuse the same const DS?

File: src/main/java/com/redislabs/redisgraph/impl/JRedisGraphTransaction.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+    /**
+     * Invokes stored procedure with arguments, in multi/exec context
+     * @param graphId a graph to perform the query on
+     * @param procedure procedure name to invoke
+     * @param args procedure arguments
+     * @return response with result set with the procedure data
+     */
+    public Response<ResultSet> graphCallProcedure(String graphId, String procedure, List<String> args  ){
+        return graphCallProcedure(graphId, procedure, args, new HashMap<>());
```
Comment (gkorland) [inline src/main/java/com/redislabs/redisgraph/impl/JRedisGraphTransaction.java]: same here HashMap(0)?

File: src/main/java/com/redislabs/redisgraph/impl/JRedisGraphTransaction.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+
+    /**
+     * Deletes the entire graph, in multi/exec context
+     * @param graphId graph to delete
+     * @return response with the deletion running time statistics
+     */
+    public Response<String> graphDeleteGraph(String graphId){
+        graphCaches.remove(graphId);
```
Comment (gkorland) [inline src/main/java/com/redislabs/redisgraph/impl/JRedisGraphTransaction.java]: Shouldn't the remove be after the delete command return?

File: src/test/java/com/redislabs/redisgraph/RedisGraphAPITest.java [hunk 6]
```diff
@@ -494,4 +496,111 @@ public void testEscapedQuery() {
         Assert.assertNotNull(api.query("social", "MATCH (n) where n.s1=%s and n.s2=%s RETURN n", "S\"\'", "S\\'\\\""));
         Assert.assertNotNull(api.query("social", "MATCH (n) where n.s1='S\"\\'' RETURN n"));
     }
+
```
Comment (gkorland) [inline src/test/java/com/redislabs/redisgraph/RedisGraphAPITest.java]: please add another test with two transactions and watch

File: pom.xml [hunk 7]
```diff
@@ -46,7 +46,14 @@
 			<distribution>repo</distribution>
 		</license>
 	</licenses>
-	
+
+	<repositories>
+		<repository>
+			<id>snapshots-repo</id>
+			<url>https://oss.sonatype.org/content/repositories/snapshots</url>
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply pom.xml]: removed

File: src/main/java/com/redislabs/redisgraph/impl/JRedisGraphTransaction.java (hunk 1 above)
Comment (DvirDukhan) [reply src/main/java/com/redislabs/redisgraph/impl/JRedisGraphTransaction.java]: fixed

File: src/main/java/com/redislabs/redisgraph/impl/JRedisGraphTransaction.java (hunk 2 above)
Comment (DvirDukhan) [reply src/main/java/com/redislabs/redisgraph/impl/JRedisGraphTransaction.java]: yes

File: src/main/java/com/redislabs/redisgraph/impl/JRedisGraphTransaction.java (hunk 3 above)
Comment (DvirDukhan) [reply src/main/java/com/redislabs/redisgraph/impl/JRedisGraphTransaction.java]: fixed

File: src/main/java/com/redislabs/redisgraph/impl/JRedisGraphTransaction.java (hunk 4 above)
Comment (DvirDukhan) [reply src/main/java/com/redislabs/redisgraph/impl/JRedisGraphTransaction.java]: fixed

File: src/main/java/com/redislabs/redisgraph/impl/JRedisGraphTransaction.java (hunk 5 above)
Comment (DvirDukhan) [reply src/main/java/com/redislabs/redisgraph/impl/JRedisGraphTransaction.java]: fixed

**Comment by K-Jo** [comment]
fixes #34

File: pom.xml [hunk 8]
```diff
@@ -62,7 +62,7 @@
 		<dependency>
 			<groupId>redis.clients</groupId>
 			<artifactId>jedis</artifactId>
-			<version>3.0.1</version>
+			<version>3.1.0-m4</version>
```
Comment (gkorland) [inline pom.xml]: 3.1.0-rc

File: src/main/java/com/redislabs/redisgraph/graph_entities/GraphEntity.java [hunk 9]
```diff
@@ -16,8 +16,8 @@
 
     //members
 
-    protected int id;
-    protected final Map<String, Property> propertyMap = new HashMap<>();
+    int id;
```
Comment (gkorland) [inline src/main/java/com/redislabs/redisgraph/graph_entities/GraphEntity.java]: do you really need it as packaged?

File: src/main/java/com/redislabs/redisgraph/impl/graph_cache/GraphCacheList.java [hunk 10]
```diff
@@ -7,17 +7,16 @@
...
+
 
 /**
  * Represents a local cache of list of strings. Holds data from a specific procedure, for a specific graph.
  */
-public class GraphCacheList {
+class GraphCacheList {
 
-    private Object mutex = new Object();
+    private final Object mutex = new Object();
```
Comment (gkorland) [inline src/main/java/com/redislabs/redisgraph/impl/graph_cache/GraphCacheList.java]: why do you need this mutex? can't you just synchronized on data?

File: pom.xml (hunk 8 above)
Comment (DvirDukhan) [reply pom.xml]: fixed

File: src/main/java/com/redislabs/redisgraph/graph_entities/GraphEntity.java (hunk 9 above)
Comment (DvirDukhan) [reply src/main/java/com/redislabs/redisgraph/graph_entities/GraphEntity.java]: fixed

File: src/main/java/com/redislabs/redisgraph/impl/graph_cache/GraphCacheList.java (hunk 10 above)
Comment (DvirDukhan) [reply src/main/java/com/redislabs/redisgraph/impl/graph_cache/GraphCacheList.java]: changed

File: src/test/java/com/redislabs/redisgraph/RedisGraphAPITest.java [hunk 11]
```diff
@@ -33,11 +36,14 @@ public void createApi(){
     public void deleteGraph() {
 
         api.deleteGraph("social");
-        api.close();
+        try {
+            api.close();
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (gkorland) [inline src/test/java/com/redislabs/redisgraph/RedisGraphAPITest.java]: should fail test

File: src/main/java/com/redislabs/redisgraph/impl/api/ContextedRedisGraph.java [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+
+    /**
+     * Sends the query over the instance only connection
+     * @param graphId graph to be queried
+     * @param preparedQuery prepared query
+     * @return Result set with the query answer
+     */
+    @Override
+    protected ResultSet sendQuery(String graphId, String preparedQuery) {
+        Jedis conn = getConnection();
```
Comment (gkorland) [inline src/main/java/com/redislabs/redisgraph/impl/api/ContextedRedisGraph.java]: try resource block?

File: src/main/java/com/redislabs/redisgraph/impl/api/ContextedRedisGraph.java [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    /**
+     * Deletes the entire graph
+     * @param graphId graph to delete
+     * @return delete running time statistics
+     */
+    @Override
+    public String deleteGraph(String graphId) {
+        Object response = getConnection().sendCommand(RedisGraphCommand.DELETE, graphId);
```
Comment (gkorland) [inline src/main/java/com/redislabs/redisgraph/impl/api/ContextedRedisGraph.java]: try resource block?

File: src/main/java/com/redislabs/redisgraph/impl/api/ContextedRedisGraph.java [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+    /**
+     * Sends the query over the instance only connection
+     * @param graphId graph to be queried
+     * @param preparedQuery prepared query
+     * @return Result set with the query answer
+     */
+    @Override
+    protected ResultSet sendQuery(String graphId, String preparedQuery) {
+        Jedis conn = getConnection();
+        List<Object> rawResponse = (List<Object>) conn.sendCommand(RedisGraphCommand.QUERY, graphId, preparedQuery, "--COMPACT");
```
Comment (gkorland) [inline src/main/java/com/redislabs/redisgraph/impl/api/ContextedRedisGraph.java]: "--COMPACT" const?

File: src/test/java/com/redislabs/redisgraph/RedisGraphAPITest.java (hunk 11 above)
Comment (DvirDukhan) [reply src/test/java/com/redislabs/redisgraph/RedisGraphAPITest.java]: `@After` annotation

File: src/main/java/com/redislabs/redisgraph/impl/api/ContextedRedisGraph.java (hunk 12 above)
Comment (DvirDukhan) [reply src/main/java/com/redislabs/redisgraph/impl/api/ContextedRedisGraph.java]: just try, no need to close resource at the end of the scope

File: src/main/java/com/redislabs/redisgraph/impl/api/ContextedRedisGraph.java (hunk 13 above)
Comment (DvirDukhan) [reply src/main/java/com/redislabs/redisgraph/impl/api/ContextedRedisGraph.java]: same as above

File: src/main/java/com/redislabs/redisgraph/impl/api/ContextedRedisGraph.java (hunk 14 above)
Comment (DvirDukhan) [reply src/main/java/com/redislabs/redisgraph/impl/api/ContextedRedisGraph.java]: fixed

File: src/main/java/com/redislabs/redisgraph/impl/api/ContextedRedisGraph.java [hunk 15]
```diff
@@ -45,7 +46,14 @@ protected Jedis getConnection() {
...
-        List<Object> rawResponse = (List<Object>) conn.sendCommand(RedisGraphCommand.QUERY, graphId, preparedQuery, "--COMPACT");
+        List<Object> rawResponse;
+        try {
+            rawResponse = (List<Object>) conn.sendCommand(RedisGraphCommand.QUERY, graphId, preparedQuery, Utils.compactString);
+        }
+        catch (Exception e) {
+            conn.close();
+            throw e;
+        }
         return new ResultSetImpl(rawResponse, this, graphId, caches.getGraphCache(graphId));
```
Comment (gkorland) [inline src/main/java/com/redislabs/redisgraph/impl/api/ContextedRedisGraph.java]: move into the catch block

**Pull request merged**